### ISSUE TITLE
update doc contribution page re sphinxext-rediraffe

### ISSUE THREAD

**Issue by phillxnet** [rockstor/rockstor-doc#296]
We no longer required the use of the sphinxcontrib-fulltoc but do now required sphinxext-rediraffe as from the merge of #290.

It is proposed that all we need, at a minimum, is to substitute our existing:
```
[you@your_laptop /path/to/rockstor-doc]# sudo pip3 install --user sphinxcontrib-fulltoc
```
for
```
[you@your_laptop /path/to/rockstor-doc]# sudo pip3 install --user sphinxext-rediraffe
```
Along with the necessary changes to the docker based instructins.

N.B. for reference on Python version compatibility we have: https://pypi.org/project/sphinxext-rediraffe/
where we have:
```
Requires: Python >=3.6
```
So it is further proposed that we move to Python3 only instructions.

**Comment by FroggyFlox** [comment]
I had a look at this and I believe we could benefit from re-arranging the content in a way that would be more clear and help the reader find the information they need quickly. As an illustration, here's our current structure:

1. Making changes
1.1 Installing Sphinx
1.2 Using Sphinx as a Docker container
1.3 Generating html files

This structure is unfortunately not the most helpful (and sub-optimal) and does not actually reflect all the steps of the process that are detailed. Indeed, the current flow of information could be described as follows:
- setup local rockstor repo
- add new changes and commit them
- installing Sphinx
- push changes to origin
 
I personally think we could opt for a better organization to keep things clearer and easier for the user to find the sought after information. We could thus go with something along the following lines (the following does not necessarily represent section headers):
1. Setup
1.1 Clone repo
1.2 Install Sphinx and sphinxext-rediraffe
2. Making changes and testing them
2.2 Guidelines (keep non-technical users in mind, ...)
2.3 Make html
2.4 Test for redirects and Create redirects if applicable
3. Submit changes
3.1 Commit
3.2 Push
3.3 Create pull request

@phillxnet, what do you think?

**Comment by phillxnet** [comment]
@FroggyFlox 
> @phillxnet, what do you think?

I think that's an excellent idea. The clearer and cleaner we can make things the better all around. We have only really tweaked this section for some time now, it has gotten a little flabby. 

We might also consider including our proposed, and in need of sphinxext-rediraffe addition, Dockerfile within the repo itself. That way we can reference it's use directly from the cloned repo. Avoiding the requirement for folks to create it. Just a thought. Might make for easier instantiation of the Sphinx setup. Not sure quite how/where-best we hide this file, thought it's likely entirely possible.

Thanks for the consideration on this one and for assigning yourself. Should definitely help with getting folks on-board with doc edits/updates/clean-ups in the future.

**Comment by FroggyFlox** [comment]
Sounds good, I'll go ahead with that plan, then.


> We might also consider including our proposed, and in need of sphinxext-rediraffe addition, Dockerfile within the repo itself. That way we can reference it's use directly from the cloned repo. Avoiding the requirement for folks to create it. Just a thought. Might make for easier instantiation of the Sphinx setup. Not sure quite how/where-best we hide this file, thought it's likely entirely possible.

That's a good idea too. It should be fairly simple so we might as well implement that now. I believe we should be able to put that Dockerfile in a `./docker/` subdirectory in the root of this repo... or we could create a repo just for that but that seems overkill. We could also host it in the Github container registry ([Github packages](https://docs.github.com/en/packages/learn-github-packages/introduction-to-github-packages)) rather than Docker hub as the former seems to be free for public repositories, whereas the latter now has limitations for any of their images on the free tier. I'll give that a try on my fork and see how it goes.

**Comment by phillxnet** [comment]
@FroggyFlox 
Re:
> I believe we should be able to put that Dockerfile in a `./docker/` subdirectory in the root of this repo..
That sounds like the way to go.

> or we could create a repo just for that but that seems overkill.

Agreed this would be overkill.

> We could also host it in the Github container registry ([Github packages](https://docs.github.com/en/packages/learn-github-packages/introduction-to-github-packages)) rather than Docker hub ... I'll give that a try on my fork and see how it goes.

That would be great, thanks.

**Comment by FroggyFlox** [comment]
@phillxnet,
As discussed above, I tried to create a docker image and publish it to the Github container registry and it went rather smoothly. I created a dedicated test repo (I will delete it once I'm done with it) for it and it seems to work well. To build the docs, for instance, one simply has to run:
```shell
 docker run --rm -v $PWD:/docs ghcr.io/froggyflox/rockstor-doc-test:main make html
```
Of course, the final one will have a better name, which I believe we should be able to set as we want. It could thus be `rockstor/rockstor-sphinx` to follow the idea first coined by @20design when they documented the creation/use of a docker image for this.

Note that it uses a Github Action to build and push this image to the Github container registry; I've set that action to be triggered only manually (one would need to go to Github.com, Actions tab, and then click on Run that workflow) as I do not foresee a need to have that triggered automatically by some event.

Let me know what you think.

**Comment by phillxnet** [comment]
@FroggyFlox  Well done. This sounds just the ticket.
Thanks for looking into and testing this docker arrangement. Looking forward to the pull request.

**Comment by FroggyFlox** [comment]
> Note also that it uses buildx by default so we can also adapt this workflow to create a Docker image compatible with the arm64 architecture if we want (at least it seems possible).

It actually seems more straightforward than I thought and I now have adapted the Github workflow to create a multiarch image (compatible with both amd64 and arm64). I haven't time to test it on arm64 yet, but the amd64 one seems to work fine. Not an essential feature but nice to have!

**Comment by FroggyFlox** [comment]
As usual, I spoke too soon as even though the Github workflow completed successfully and built an image said to be arm64-compatible, it was not working when testing it on a Raspberry Pi4. After a few trials and errors, I figured out that there was nothing wrong with the Github workflow, but that the problem lies with the fact that the `sphinxdoc/sphinx` image we used as a base (as per Sphinx's recommendation) is not arm64-compatible. 
A workaround is to bypass the `sphinxdoc/sphinx` image and use [their Dockerfile](https://github.com/sphinx-doc/docker/blob/master/base/Dockerfile) that is itself using the `python:slim` image as the latter is arm64-compatible. In other words, if we copy `sphinxdoc/sphinx` Dockerfile and add our "customization", _i.e._ installing `git` and `sphinxext-rediraffe`, then we can build a working image for both amd64 and arm64 platforms.
@phillxnet , I'm personally not a fan of that approach, though, especially given the fact that having an arm64 image for this is not a must. We could thus just keep it simple with an amd64-only image... for the moment at least.

**Comment by phillxnet** [comment]
@FroggyFlox 

> We could thus just keep it simple with an amd64-only image... for the moment at least.

Sound good. Also, good to know the sphinxdoc/sphinx image is not ARM64 compatible. Nice find. It might be worth a request for them to extend their architecture compatibility actually. More and more image maintainers are seeing to such things so they may well be receptive.

**Comment by FroggyFlox** [comment]
> It might be worth a request for them to extend their architecture compatibility actually.

Agreed... It crossed my mind but the Issue tracker is not enabled on the sphinxdoc/sphinx docker image Github repo so I'll try to see what is their preferred method of contact.
